fix emitting of more windows than requested

MartinNowak opened the pull request:
- fixup for #12
- must not eagerly emit a new window when the predicate triggers
- fixed by emitting the window on the next requested value (falling edge-trigger)
- for Mode.BEFORE it is necessary to save the current value until a
  new window gets requested (1 element look-ahead)
- backpressure tests didn't fully work as with `flatMapSingle(toList)` only
  values for completed inner windows where tested, test emitted counts instead

The level of complexity is quite a mouthful, despite having read your elaborated warning ;).
http://akarnokd.blogspot.de/2015/05/pitfalls-of-operator-implementations.html
http://akarnokd.blogspot.de/2015/05/pitfalls-of-operator-implementations_14.html

akarnokd commented on src/main/java/hu/akarnokd/rxjava2/operators/FlowableWindowPredicate.java:
```diff
@@ -92,6 +92,8 @@ protected void subscribeActual(Subscriber<? super Flowable<T>> s) {
 
         UnicastProcessor<T> window;
 
+        T pending; // look-ahead element for Mode.BEFORE
```
There was a similar problem with `scan` that caused unwanted element delay. You may need to have full queue-drain on the window-emission side to reliably solve the problem, similar to `groupBy`, because we can't know if the next element will trigger a new window or not.

MartinNowak replied:
Ah good tip to look at GroupBy.
We have a much simpler problem here though, as there is only ever one active window.
When finding the element that toggles the predicate, no element is requested from upstream, so at that point the number of upstream requests is equal to the number of requested windows downstream. Because of this we can safely emit a new window in the onNext call.
The only tricky thing is that the BEFORE split mode (windowUntil) will emit 2 elements to the newly created window at once.
We could try to check the requested number of downstream windows and emit the window immediately if allowed, but we'd still have to defer the emission of the element when no more windows were requested.

akarnokd replied:
If there is an element held in `pending` due to the case you described, a request from downstream should trigger the emission of a window immediately since the upstream before the operator may not emit the next item in a long time or ever. For this you need the full queue-drain trampolining found in many operators.

MartinNowak commented:
I've implemented a 1-element drain-queue using AtomicReference and added a test case for the delayed upstream.
Also added some tests for when the predicate matches the first element.

MartinNowak commented:
Retested after update to JDK8. Anything left to do?

akarnokd commented:
Thanks.
